### DISCUSSION TITLE
refute_has doesn't block (for async code)

### DISCUSSION THREAD

**Discussion by jonleighton** [keathley/wallaby#587] [Q&A]
I wrote a test like this:

``` elixir
session
|> refute_has(css(".my-element"))
```

The element is becomes hidden via JS executing on the page.

My expectation was that if the element was present when `refute_has` first runs, it would wait for the async timeout and keep testing the DOM before raising an error. This turns out not to be true, which lead to flakey test failures.

We can do this instead:

``` elixir
session
|> assert_has(css(".my-element", count: 0))
```

But I find the behaviour of `refute_has` surprising. Is this deliberate? If not, perhaps it can be changed?

**Comment by mhanberg** [comment]
I believe it is deliberate, but it was written before my time. I think there has been discussion to deprecate the function entirely in favor of the method you provided, but I believe that discussion is all it's been.

**Reply by jonleighton** [reply]
Thanks for the response and moving this to a discussion. I didn't even realise GH had this feature now! :grimacing: 

One advantage for me of keeping `refute_has` but changing the behaviour is that it would allow us to reuse the selector for both `assert_has` and `refute_has`. On the code base I'm working on, we have "page objects" (modules), so our test code looks like this:

``` elixir
session
|> refute_has(MyPage.some_element())
```

For now we've ended up just passing the `count` option through:

``` elixir
session
|> assert_has(MyPage.some_element(count: 0))
```

But I don't love the fact that that exposes more of a lower level detail to the higher level code.

**Reply by mhanberg** [reply]
You can also reuse the query but modify the count like

```elixir
session
|> assert_has(MyPage.some_element() |> Query.count(0))
```

**Reply by jonleighton** [reply]
We're introducing an `assert_missing` function for our tests:

``` elixir
  def assert_missing(parent, query) do
    assert_has(parent, query |> Query.count(0))
  end
```

Perhaps that's something that could be useful to add to Wallaby core?

It makes the tests read a bit nicer and means we don't need to think so much about whether or not async behaviour might trip us up.

**Reply by marcandre** [reply]
I also believe that the current `refute_has` should be changed:
1) It is not useful that it doesn't retry when an element is currently visible. All tests should be willing to spend `max_wait_time` before failing.
2) It currently systematically waits until `max_wait_time` before being a success. Seems clearly undesirable, and for high `max_wait_time` makes things unbearable. 
3) I believe that it will succeed even if you pass an invalid query (e.g. `refute_has(Query.css("#test", minimum: 5, maximum: 3))`)

**Comment by marcandre** [comment]
Would a PR fixing this be accepted? @mhanberg

**Reply by mhanberg** [reply]
Yes, I think it would be best to approach it like this https://github.com/elixir-wallaby/wallaby/discussions/587#discussioncomment-876303. 

Keeping the current naming, tho.

**Reply by marcandre** [reply]
Great. I hope to be able to propose something soonish (travelling right now).

**Reply by vanderhoop** [reply]
Hey @mhanberg + @marcandre Seems like this convo is a bit stale. I've got some time, and just ran into a need for this (for not the first time 😀). Will try to take it on this weekend unless I hear otherwise.

**Reply by marcandre** [reply]
Go for it @vanderhoop . I'm no longer using `wallaby` (we've opted to use `cypress` instead)
